### PR TITLE
Prevent multiple editors on a submission

### LINKED ISSUE #1076: Implement locking for metadata submissions
Part of #824 

Currently, we don't have a robust set of role based access controls for metadata submissions to the NMDC Submission Portal. Part of implementing these permissions will be introducing some sort of record locking to prevent concurrent edits of a submission.

Implementation of locking is NOT blocked by permissions, since all users marked as "Admins" already have access to all submissions. Therefore, testing of these locks can be done independently of the actual role based permissions.

It might be a good idea to start with this.

### PR DESCRIPTION
**Pull request by naglepuff**
Fix #1076 

This group of changes disables multiple people accessing a single submission at the same time. This prevents users getting their changes overridden by others on the rare occasion that they are editing the same submission simultaneously. It contains front end and back end changes.

## Backend

1. Data model changes: the `SubmissionMetadata` model now contains a second foreign key to the user table representing the user who currently holds the "lock" for that submission, and a timestamp representing when the last edit was made. A submission "lock" refers to the combination of these new fields.
2. An accompanying migration file is included.
3. New functions were added in `crud.py` for attempting to obtain the lock, updating the lock, and releasing the lock.
4. Existing `/api/metadata_submission` routes were updated to include lock checking
5. A new endpoint was added to release the lock for a submission.

## Frontend

@mslarae13 a full read-only mode will likely be available once #1077 is developed, which will likely replace this approach. I'm happy to discuss further at squad meetings.

1. If a user attempts to access a locked submission from the submission list, they will be shown a message that informs them that the submission is currently locked, and offer to take them back to the submission list. The name (if available, ORCID iD if not) is shown to the user.
2. If a user editing a submission navigates away from any screen that lets them edit the submission (for example, going back to the submission list, navigating away from the NMDC Submission Portal entirely, or closing the tab/window), the submission becomes unlocked.

 
image

Note for testing before this gets deployed: I believe ORCID doesn't like people having multiple accounts. To test this locally I used the ORCID sandbox environment [sandbox.orcid.org ](https://sandbox.orcid.org/)to create multiple users (using variations on mike.nagler+xyz@kitware.com). Reach out for more help setting this up.

### CODE REVIEW

**Comment by mslarae13** [comment]
looks great Mike! Maybe we can demo the next time we meet for UFO.

File: web/src/views/SubmissionPortal/StepperView.vue [hunk 1]
```diff
@@ -1,18 +1,79 @@
...
+  },
+
 });
 </script>
 
 <template>
   <div>
     <SubmissionStepper class="mx-0" />
-    <v-container>
+    <!-- Put locked by notification/banner here -->
```
Comment (marySalvi) [inline web/src/views/SubmissionPortal/StepperView.vue:60]: Is this comment still needed?

**Review by marySalvi — CHANGES_REQUESTED** [review]

File: web/src/views/SubmissionPortal/StepperView.vue (hunk 1 above)
Comment (naglepuff) [reply web/src/views/SubmissionPortal/StepperView.vue]: Whoops. removed in f384bfc

File: nmdc_server/crud.py [hunk 2]
```diff
@@ -425,3 +426,65 @@ def update_user(db: Session, user: schemas.User) -> Optional[models.User]:
...
+            # A user can hold the lock for 30 minutes without making edits
+            if seconds_since_last_lock_update > (60 * 30):
+                submission_record.locked_by = user_record
+                submission_record.lock_updated = datetime.utcnow()
+                db.commit()
+                return True
+            else:
+                # Someone else holds the lock, but there's no timestamp
+                # Ensure that there's a timestamp on the lock.
+                submission_record.lock_updated = datetime.utcnow()
```
[the remark this answers is not in the document]
Comment (naglepuff) [reply nmdc_server/crud.py]: Good catch. I un-indented that last else clause. Which is kind of a paranoid guard against submissions that are locked with no timestamp. addb7fb

File: nmdc_server/api.py [hunk 3]
```diff
@@ -539,6 +539,7 @@ async def get_submission(
     submission = db.query(SubmissionMetadata).get(id)
     if submission is None:
         raise HTTPException(status_code=404, detail="Submission not found")
+    _ = crud.try_get_submission_lock(db, submission.id, user.id)
```
[the remark this answers is not in the document]
Comment (naglepuff) [reply nmdc_server/api.py]: I removed the assignment and reordered the checks in 2298bc5

The behavior of this endpoint is the same regardless of whether or not the user requesting has the submission locked, which is why we don't need to use the returned value.

**Review by marySalvi — APPROVED** [review]

**Pull request merged**